evaluator/builtin_time.go

```diff
@@ -673,6 +675,330 @@ func builtinFromUnixTime(args []types.Datum, _ context.Context) (d types.Datum,
+		monthName := input[:3]
+		if month, ok := monthAbbrev[monthName]; ok {
+			timeSetMonth(t, month)
+			return input[len(monthName):], true
+		}
+	}
+	return input, false
+}
+
+func monthNumeric(t *time.Time, input string) (string, bool) {
+	// TODO: this code is ugly!
+	for i := 12; i >= 0; i-- {
+		str := strconv.FormatInt(int64(i), 10)
+		if strings.HasPrefix(input, str) {
+			timeSetMonth(t, time.Month(i))
+			return input[len(str):], true
+		}
+	}
+
+	return input, false
+}
+
+// 0th 1st 2nd 3rd ...
+func dayOfMonthWithSuffix(t *time.Time, input string) (string, bool) {
+	month, remain := parseOrdinalNumbers(input)
+	if month >= 0 {
+		timeSetMonth(t, time.Month(month))
+		return remain, true
+	}
+	return input, false
+}
+
+func parseOrdinalNumbers(input string) (value int, remain string) {
+	for i, c := range input {
+		if !unicode.IsDigit(c) {
+			v, err := strconv.ParseUint(input[:i], 10, 64)
+			if err != nil {
+				return -1, input
+			}
+			value = int(v)
+			break
+		}
+	}
+	switch {
+	case strings.HasPrefix(remain, "st"):
+		if value == 1 {
+			remain = remain[2:]
+			return
+		}
+	case strings.HasPrefix(remain, "nd"):
+		if value == 2 {
+			remain = remain[2:]
+			return
+		}
+	case strings.HasPrefix(remain, "th"):
+		remain = remain[2:]
+		return
+	}
+	return -1, input
+}
+
+func timeSetYear(t *time.Time, year int) {
+	_, month, day := t.Date()
+	hour, min, sec := t.Clock()
+	nsec := t.Nanosecond()
+	loc := t.Location()
+	*t = time.Date(year, month, day, hour, min, sec, nsec, loc)
+}
+
+func timeSetMonth(t *time.Time, month time.Month) {
+	year, _, day := t.Date()
+	hour, min, sec := t.Clock()
+	nsec := t.Nanosecond()
+	loc := t.Location()
+	*t = time.Date(year, month, day, hour, min, sec, nsec, loc)
+}
+
+func timeSetDay(t *time.Time, day int) {
+	year, month, _ := t.Date()
+	hour, min, sec := t.Clock()
+	nsec := t.Nanosecond()
+	loc := t.Location()
+	*t = time.Date(year, month, day, hour, min, sec, nsec, loc)
+}
+
+func timeSetHour(t *time.Time, hour int) {
+	year, month, day := t.Date()
+	_, min, sec := t.Clock()
+	nsec := t.Nanosecond()
+	loc := t.Location()
+	*t = time.Date(year, month, day, hour, min, sec, nsec, loc)
+}
+
+func timeSetMinute(t *time.Time, min int) {
+	year, month, day := t.Date()
+	hour, _, sec := t.Clock()
+	nsec := t.Nanosecond()
+	loc := t.Location()
+	*t = time.Date(year, month, day, hour, min, sec, nsec, loc)
+}
+
+func timeSetSecond(t *time.Time, sec int) {
+	year, month, day := t.Date()
+	hour, min, _ := t.Clock()
+	nsec := t.Nanosecond()
+	loc := t.Location()
+	*t = time.Date(year, month, day, hour, min, sec, nsec, loc)
+}
+
+// See https://dev.mysql.com/doc/refman/5.5/en/date-and-time-functions.html#function_str-to-date
+func builtinStrToDate(args []types.Datum, _ context.Context) (types.Datum, error) {
+	date := args[0].GetString()
+	format := args[1].GetString()
+	var (
+		d      types.Datum
+		goTime time.Time
+	)
+	goTime = mysql.ZeroTime
+	if !strToDate(&goTime, date, format) {
+		d.SetNull()
```

Comment:
I prefer `d.SetNull()` here explicitly, it can remind us this function return NULL if strToDate fail...
depend on KindNull == 0 is not that reliable on some scenes